### PR TITLE
Matched d_a_obj_tower

### PR DESCRIPTION
**Pull request by link5669**
Resolved issue #441

### CODE REVIEW

File: src/d/actor/d_a_obj_tower.cpp [hunk 1]
```diff
@@ -4,50 +4,171 @@
...
+#include "dolphin/mtx/mtx.h"
 #include "dolphin/types.h"
+#include "f_op/f_op_actor_mng.h"
+#include "m_Do/m_Do_printf.h"
 
 /* 00000078-00000098       .text CheckCreateHeap__FP10fopAc_ac_c */
-static BOOL CheckCreateHeap(fopAc_ac_c*) {
-    /* Nonmatching */
+static BOOL CheckCreateHeap(fopAc_ac_c* i_this) {
+    ((daObjTower_c*)i_this)->CreateHeap();
```
Comment (magcius) [inline src/d/actor/d_a_obj_tower.cpp:16]: Missing return

File: src/d/actor/d_a_obj_tower.cpp [hunk 2]
```diff
@@ -4,50 +4,171 @@
...
 }
 
 /* 00000098-0000020C       .text CreateHeap__12daObjTower_cFv */
-void daObjTower_c::CreateHeap() {
-    /* Nonmatching */
+bool daObjTower_c::CreateHeap() {
+    J3DModelData* pModel = (J3DModelData*)dRes_control_c::getRes(
+        "X_tower", 4, g_dComIfG_gameInfo.mResControl.mObjectInfo, 0x40);
```
Comment (magcius) [inline src/d/actor/d_a_obj_tower.cpp:22]: dComIfG_getObjectRes

File: src/d/actor/d_a_obj_tower.cpp [hunk 3]
```diff
@@ -4,50 +4,171 @@
...
+
+    mpModel = mDoExt_J3DModel__create(pModel, 0, 0x11020203);
+
+    if (mpModel == NULL) {
+        return false;
+    } else {
+        PSMTXTrans(mDoMtx_stack_c::now, current.pos.x, current.pos.y, current.pos.z);
+        mDoMtx_YrotM(mDoMtx_stack_c::now, shape_angle.y);
+        mDoMtx_stack_c::scaleM(mScale.x, mScale.y, mScale.z);
+        PSMTXCopy(mDoMtx_stack_c::now, mMtx);
```
Comment (magcius) [inline src/d/actor/d_a_obj_tower.cpp:38]:
```suggestion
        mDoMtx_stack_c::transS(current.pos);
        mDoMtx_stack_c::YrotM(shape_angle.y);
        mDoMtx_stack_c::scaleM(mScale.x, mScale.y, mScale.z);
        mDoMtx_copy(mDoMtx_stack_c::get(), mMtx);
```

File: src/d/actor/d_a_obj_tower.cpp [hunk 4]
```diff
@@ -4,50 +4,171 @@
...
+                "X_tower", 7, g_dComIfG_gameInfo.mResControl.mObjectInfo, 0x40);
+            return !mpBgW->Set(pData, cBgW::MOVE_BG_e, &mMtx);
+        }
+    }
 }
 
 /* 0000020C-0000028C       .text CreateInit__12daObjTower_cFv */
 void daObjTower_c::CreateInit() {
-    /* Nonmatching */
+    mCullMtx = mpModel->mBaseTransformMtx;
```
Comment (magcius) [inline src/d/actor/d_a_obj_tower.cpp:54]:
```suggestion
    fopAcM_SetMtx(this, mpModel->getBaseTRMtx());
```

File: src/d/actor/d_a_obj_tower.cpp [hunk 5]
```diff
@@ -4,50 +4,171 @@
...
+    fopAcM_setCullSizeBox(this, -30000.0, -5000.0, -30000.0, 30000.0, 40000.0, 30000.0);
+    g_dComIfG_gameInfo.play.mBgS.Regist(mpBgW, this);
```
Comment (magcius) [inline src/d/actor/d_a_obj_tower.cpp:56]:
```suggestion
    dComIfGp_Bgsp()->Regist(mpBgW, this);
```

File: src/d/actor/d_a_obj_tower.cpp [hunk 6]
```diff
@@ -4,50 +4,171 @@
...
 
 /* 0000028C-0000030C       .text set_mtx__12daObjTower_cFv */
 void daObjTower_c::set_mtx() {
-    /* Nonmatching */
+    J3DModel* pJVar1;
+    pJVar1 = mpModel;
+
+    (pJVar1->mBaseScale).x = mScale.x;
+    (pJVar1->mBaseScale).y = mScale.y;
+    (pJVar1->mBaseScale).z = mScale.z;
```
Comment (magcius) [inline src/d/actor/d_a_obj_tower.cpp:68]:
```suggestion
    mpModel->setBaseScale(mScale);
```

File: src/d/actor/d_a_obj_tower.cpp [hunk 7]
```diff
@@ -4,50 +4,171 @@
...
+
+    PSMTXTrans(mDoMtx_stack_c::now, current.pos.x, current.pos.y, current.pos.z);
+    mDoMtx_YrotM(mDoMtx_stack_c::now, current.angle.y);
+    PSMTXCopy(mDoMtx_stack_c::now, mpModel->mBaseTransformMtx);
```
Comment (magcius) [inline src/d/actor/d_a_obj_tower.cpp:72]: Use the mDoMtx_stack_c API, as above.

File: src/d/actor/d_a_obj_tower.cpp [hunk 8]
```diff
@@ -4,50 +4,171 @@
...
 }
 
 /* 0000030C-000003DC       .text daObjTower_Create__FPv */
-static s32 daObjTower_Create(void*) {
-    /* Nonmatching */
+static cPhs__Step daObjTower_Create(void* i_this) {
+    cPhs__Step PVar3;
```
Comment (magcius) [inline src/d/actor/d_a_obj_tower.cpp:77]: As seen in the debug headers, daObjTower_Create calls an inline `daObjTower::_create`, which should help improve the casts here. The implementation of this method should just be:

```cpp
static s32 daObjTower_Create(void* i_this) {
    return ((daObjTower_c*)i_this)->_create();
}
```

with the rest of this method pulled into the _create method. This also goes for `_draw`, `_execute`, and `_delete`.

File: src/d/actor/d_a_obj_tower.cpp [hunk 9]
```diff
@@ -4,50 +4,171 @@
...
+
+    fopAcM_SetupActor(((daObjTower_c*)i_this), daObjTower_c);
+
+    ((daObjTower_c*)i_this)->field_0x2d0 = 0;
+
+    if (!g_dComIfG_gameInfo.save.mSavedata.mEvent.isEventBit(0x1e40)) {
```
Comment (magcius) [inline src/d/actor/d_a_obj_tower.cpp:83]: dComIfGs_isEventBit

File: src/d/actor/d_a_obj_tower.cpp [hunk 10]
```diff
@@ -4,50 +4,171 @@
...
+    return PVar3;
 }
 
 /* 000003DC-00000444       .text daObjTower_Delete__FPv */
-static BOOL daObjTower_Delete(void*) {
-    /* Nonmatching */
+static BOOL daObjTower_Delete(void* i_this) {
+    if (((daObjTower_c*)(i_this))->field_0x2d0 == 1) {
+        ((cBgS*)&g_dComIfG_gameInfo.play.mBgS.m_chk_element[0].m_bgw_base_ptr)
+            ->Release((cBgW*)((daObjTower_c*)i_this)->mpBgW);
```
Comment (magcius) [inline src/d/actor/d_a_obj_tower.cpp:102]:
```suggestion
        dComIfGp_Bgsp()->Release(mpBgW);
```

File: src/d/actor/d_a_obj_tower.cpp [hunk 11]
```diff
@@ -4,50 +4,171 @@
...
+    }
+
+    if (((daObjTower_c*)i_this)->mBase.mCreateResult != 3) {
```
Comment (magcius) [inline src/d/actor/d_a_obj_tower.cpp:105]:
```suggestion
    if (fpcM_CreateResult(i_this) != cPhs_UNK3_e) {
```

File: src/d/actor/d_a_obj_tower.cpp [hunk 12]
```diff
@@ -4,50 +4,171 @@
...
-static BOOL daObjTower_Draw(void*) {
-    /* Nonmatching */
+static BOOL daObjTower_Draw(void* i_this) {
+    g_env_light.settingTevStruct(1, &((daObjTower_c*)i_this)->current.pos,
+                                 &((daObjTower_c*)i_this)->mTevStr);
+    g_env_light.setLightTevColorType(((daObjTower_c*)i_this)->mpModel,
+                                     &((daObjTower_c*)i_this)->mTevStr);
+
+    j3dSys.mDrawBuffer[0] = g_dComIfG_gameInfo.drawlist.mpOpaListBG;
+    j3dSys.mDrawBuffer[1] = g_dComIfG_gameInfo.drawlist.mpXluListBG;
```
Comment (magcius) [inline src/d/actor/d_a_obj_tower.cpp:120]:
```suggestion
    dComIfGd_setListBG();
```

File: src/d/actor/d_a_obj_tower.cpp [hunk 13]
```diff
@@ -4,50 +4,171 @@
...
+
+    mDoExt_modelUpdateDL(((daObjTower_c*)i_this)->mpModel);
+
+    j3dSys.mDrawBuffer[0] = g_dComIfG_gameInfo.drawlist.mpOpaList;
+    j3dSys.mDrawBuffer[1] = g_dComIfG_gameInfo.drawlist.mpXluList;
```
Comment (magcius) [inline src/d/actor/d_a_obj_tower.cpp:125]:
```suggestion
    dComIfGd_setList();
```

File: src/d/actor/d_a_obj_tower.cpp [hunk 14]
```diff
@@ -4,50 +4,171 @@
...
+static BOOL daObjTower_Execute(void* i_this) {
+    u8 idx = ((daObjTower_c*)i_this)->mDemoActorId;
+
+    if (idx != 0) {
+        dDemo_actor_c* pdVar1 = (g_dComIfG_gameInfo.play.mDemo)->mDemoObj.getActor(idx);
+        if ((pdVar1 != 0) && (pdVar1->mFlags & 2)) {
+            ((daObjTower_c*)i_this)->current.pos.x = (pdVar1->mTranslation).x;
+            ((daObjTower_c*)i_this)->current.pos.y = (pdVar1->mTranslation).y;
+            ((daObjTower_c*)i_this)->current.pos.z = (pdVar1->mTranslation).z;
+        }
```
Comment (magcius) [inline src/d/actor/d_a_obj_tower.cpp:140]:
```suggestion
        dDemo_actor_c* pdVar1 = dComIfGp_demo_getActor(mDemoActorId);
        if ((pdVar1 != 0) && (pdVar1->checkEnable(0x02)) {
            i_this->current.pos = pdVar1->mTranslation;
        }
```

**Review by magcius — CHANGES_REQUESTED** [review]
Nice! Mostly just missing some inlines here.

File: src/d/actor/d_a_obj_tower.cpp [hunk 15]
```diff
@@ -4,50 +4,164 @@
...
 
 /* 000003DC-00000444       .text daObjTower_Delete__FPv */
-static BOOL daObjTower_Delete(void*) {
-    /* Nonmatching */
+static BOOL daObjTower_Delete(void* i_this) {
+    return ((daObjTower_c*)i_this)->_delete();
+}
+
+BOOL daObjTower_c::_draw() {
+    g_env_light.settingTevStruct(1, &this->current.pos, &this->mTevStr);
```
Comment (LagoLunatic) [inline src/d/actor/d_a_obj_tower.cpp:107]:
```suggestion
    g_env_light.settingTevStruct(TEV_TYPE_BG0, &this->current.pos, &this->mTevStr);
```

File: src/d/actor/d_a_obj_tower.cpp [hunk 16]
```diff
@@ -4,50 +4,164 @@
...
+    mDoMtx_stack_c::YrotM(current.angle.y);
+    mDoMtx_copy(mDoMtx_stack_c::get(), mpModel->mBaseTransformMtx);
 }
 
+cPhs__Step daObjTower_c::_create() {
+    cPhs__Step PVar3;
+
+    fopAcM_SetupActor(this, daObjTower_c);
+
+    this->field_0x2d0 = 0;
```
Comment (LagoLunatic) [inline src/d/actor/d_a_obj_tower.cpp:67]:
```suggestion
    field_0x2d0 = 0;
```
you can remove the `this->` from all these C++-style member functions. only `i_this->` is needed in C-style functions.

File: src/d/actor/d_a_obj_tower.cpp [hunk 17]
```diff
@@ -4,50 +4,164 @@
...
+    set_mtx();
 }
 
 /* 0000028C-0000030C       .text set_mtx__12daObjTower_cFv */
 void daObjTower_c::set_mtx() {
-    /* Nonmatching */
+    mpModel->setBaseScale(mScale);
+    mDoMtx_stack_c::transS(current.pos);
+    mDoMtx_stack_c::YrotM(current.angle.y);
+    mDoMtx_copy(mDoMtx_stack_c::get(), mpModel->mBaseTransformMtx);
```
Comment (LagoLunatic) [inline src/d/actor/d_a_obj_tower.cpp:59]:
```suggestion
    mpModel->setBaseTRMtx(mDoMtx_stack_c::get());
```

File: src/d/actor/d_a_obj_tower.cpp [hunk 18]
```diff
@@ -4,50 +4,164 @@
...
+    cBgD_t* pData = (cBgD_t*)dComIfG_getObjectRes("X_tower", 7);
+
+    return mpBgW->Set(pData, cBgW::MOVE_BG_e, &mMtx) ? FALSE : TRUE;
 }
 
 /* 0000020C-0000028C       .text CreateInit__12daObjTower_cFv */
 void daObjTower_c::CreateInit() {
-    /* Nonmatching */
+    fopAcM_SetMtx(this, mpModel->getBaseTRMtx());
+    fopAcM_setCullSizeBox(this, -30000.0, -5000.0, -30000.0, 30000.0, 40000.0, 30000.0);
```
Comment (magcius) [inline src/d/actor/d_a_obj_tower.cpp:48]:
```suggestion
    fopAcM_setCullSizeBox(this, -30000.0f, -5000.0f, -30000.0f, 30000.0f, 40000.0f, 30000.0f);
```

**Review by magcius — COMMENTED** [review]
One more I missed, sorry.

**Pull request merged**